File: FrontEnd/lib/shared/models/notification.dart
/// Modèle représentant une notification de l'application.
class AppNotification {
  const AppNotification({
    required this.id,
    this.type,
    this.title,
    this.body,
    this.dataPayload,
    this.isRead = false,
    this.readAt,
    this.createdAt,
  });

  /// Identifiant unique de la notification.
  final String id;
  /// Type de notification (pantry, budget, mealPlan, etc.).
  final String? type;
  /// Titre de la notification.
  final String? title;
  /// Corps du message.
  final String? body;
  /// Payload de données supplémentaires.
  final String? dataPayload;
  /// Indique si la notification a été lue.
  final bool isRead;
  /// Date de lecture.
  final String? readAt;
  /// Date de création.
  final String? createdAt;

  /// Crée une [AppNotification] depuis une map JSON.
  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id']?.toString() ?? '',
      type: json['type'] as String?,
      title: json['title'] as String?,
      body: json['body'] as String?,
      dataPayload: json['data'] as String?,
      isRead: json['read'] as bool? ?? false,
      readAt: json['readAt'] as String?,
      createdAt: json['createdAt'] as String?,
    );
  }

  /// Retourne une copie avec les champs modifiés.
  AppNotification copyWith({
    String? id,
    String? type,
    String? title,
    String? body,
    String? dataPayload,
    bool? isRead,
    String? readAt,
    String? createdAt,
  }) {
    return AppNotification(
      id: id ?? this.id,
      type: type ?? this.type,
      title: title ?? this.title,
      body: body ?? this.body,
      dataPayload: dataPayload ?? this.dataPayload,
      isRead: isRead ?? this.isRead,
      readAt: readAt ?? this.readAt,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  /// Convertit cette [AppNotification] en map JSON.
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'title': title,
      'body': body,
      'data': dataPayload,
    };
  }
}

File: BackEnd/src/main/java/com/platepilote/platepilote/notification/domain/entity/Notification.java
package com.platepilote.platepilote.notification.domain.entity;

import com.platepilote.platepilote.common.kernel.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.UUID;

/**
 * Entité représentant une notification destinée à un utilisateur.
 * <p>
 * Les notifications peuvent être de différents types (PANTRY_ALERT,
 * MEAL_PLAN_REMINDER, RECIPE_RECOMMENDATION, etc.) et contiennent
 * un titre, un corps de message et des données JSON optionnelles.
 * Hérite de BaseEntity pour le suivi des dates et le soft-delete.
 */
@Entity
@Table(name = "notifications")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Notification extends BaseEntity {

    /** Identifiant de l'utilisateur destinataire. */
    @Column(name = "user_id", nullable = false)
    private UUID userId;

    /** Type de notification (PANTRY_ALERT, MEAL_PLAN_REMINDER, etc.). */
    @Column(nullable = false)
    private String type;

    /** Titre de la notification. */
    @Column(nullable = false)
    private String title;

    /** Corps du message de la notification. */
    @Column(columnDefinition = "TEXT")
    private String body;

    /** Données JSON additionnelles associées à la notification. */
    @Column(columnDefinition = "jsonb")
    @JdbcTypeCode(SqlTypes.JSON)
    private String data;

    /** Indique si la notification a été lue. */
    @Column(nullable = false)
    @Builder.Default
    private Boolean read = false;

    /** Date à laquelle la notification a été lue. */
    @Column(name = "read_at")
    private Instant readAt;
}
